web/src/filter.ts

```diff
@@ -726,6 +726,53 @@ export class Filter {
         return true;
     }
 
+    static adjusted_terms_if_moved(raw_terms: NarrowTerm[], message: Message): NarrowTerm[] | null {
+        if (message.type !== "stream") {
+            return null;
+        }
+
+        assert(typeof message.display_recipient === "string");
+        assert(typeof message.topic === "string");
+
+        const adjusted_terms = [];
+        let terms_changed = false;
+
+        for (const term of raw_terms) {
+            const adjusted_term = {...term};
+            if (
+                Filter.canonicalize_operator(term.operator) === "channel" &&
+                !util.lower_same(term.operand, message.display_recipient)
+            ) {
+                adjusted_term.operand = message.display_recipient;
+                terms_changed = true;
+            }
+
+            if (
+                Filter.canonicalize_operator(term.operator) === "topic" &&
+                !util.lower_same(term.operand, message.topic)
+            ) {
+                adjusted_term.operand = message.topic;
+                terms_changed = true;
+            }
+
+            adjusted_terms.push(adjusted_term);
+        }
```

Comment:
This algorithm seems wrong in the event that the operator is just `/#narrow/with/{message_id}`, which we should definitely support to find the conversation containing that channel... Since it won't insert additional terms if none were present.

Reply:
updated in https://github.com/zulip/zulip/pull/30114#discussion_r1660233503